include/shelf-pack.hpp

```diff
@@ -57,22 +86,25 @@ class Shelf {
 
     /**
      * Allocate a single bin into the shelf.
+     * Memory to store the bin is dynamically allocated on the heap.
      *
      * @private
+     * @param    {int32_t}  id    Unique bin identifier, pass -1 to generate a new one
      * @param    {int32_t}  w     Width of the bin to allocate
      * @param    {int32_t}  h     Height of the bin to allocate
-     * @returns  {optional<Bin>}  `Bin` struct with `x`, `y`, `w`, `h` members
+     * @returns  {Bin*}     `Bin` with `id`, `x`, `y`, `w`, `h` members
+     *
      * @example
-     * optional<Bin> result = shelf.alloc(12, 16);
+     * Bin* result = shelf.alloc(-1, 12, 16);
      */
-    optional<Bin> alloc(int32_t w, int32_t h) {
+    Bin* alloc(int32_t id, int32_t w, int32_t h) {
         if (w > wfree_ || h > h_) {
-            return nullopt;
+            return NULL;
         }
         int32_t x = x_;
         x_ += w;
         wfree_ -= w;
-        return Bin(w, h, x, y_);
+        return new Bin(id, w, h, w, h_, x, y_);
```

Comment:
Thanks for the feedback!

> Use std::unique_ptr rather than raw pointers (only internally -- Bin* still used in the external API). This is awkward when it comes to moving bins between the free list and map.

Maybe I can avoid moving them, I'll think about that.

> Use Bin by value in the collections (again only internally -- the external API still hands out Bin* pointers into the map). This is a clean implementation internally, but crashes the tests because they assume that bin pointers remain valid even when the reference count goes to zero and the bin moves to the free list. Is that part of the public API contract?

It's probably not necessary for the pointers to remain valid externally after they are marked as free.  This is probably the cleanest solution.

> A third, more invasive option, which I want to think about more in the context of fixing mapbox/mapbox-gl-native#2202, is to use std::shared_ptr<Bin>. The idea would be to lean on shared_ptr's automatic reference counting rather than having manual ref and unref methods in ShelfPack.

I considered this, but I don't think I want the Bins destroyed when the refcount goes to zero.  They represent real space in the sprite that can be reclaimed.